src/Microsoft.OData.Core/MultipartMixed/ODataMultipartMixedBatchReaderStream.cs

```diff
@@ -0,0 +1,679 @@
+        private IEnumerable<string> CurrentBoundaries
+        {
+            get
+            {
+                if (this.changesetBoundary != null)
+                {
+                    yield return this.changesetBoundary;
+                }
+
+                yield return this.batchBoundary;
+            }
+        }
+
+        /// <summary>
+        /// Resets the changeset boundary at the end of the changeset.
+        /// </summary>
+        internal void ResetChangeSetBoundary()
+        {
+            Debug.Assert(this.changesetBoundary != null, "Only valid to reset a changeset boundary if one exists.");
+            this.changesetBoundary = null;
+            this.changesetEncoding = null;
+        }
+
+        /// <summary>
+        /// Skips all the data in the stream until a boundary is found.
+        /// </summary>
+        /// <param name="isEndBoundary">true if the boundary that was found is an end boundary; otherwise false.</param>
+        /// <param name="isParentBoundary">true if the detected boundary is a parent boundary (i.e., the expected boundary is missing).</param>
+        /// <returns>true if a boundary was found; otherwise false.</returns>
+        internal bool SkipToBoundary(out bool isEndBoundary, out bool isParentBoundary)
+        {
+            // Ensure we have a batch encoding; if not detect it on the first read/skip.
+            this.EnsureBatchEncoding(this.rawInputContext.Stream);
+
+            ODataBatchReaderStreamScanResult scanResult = ODataBatchReaderStreamScanResult.NoMatch;
+            while (scanResult != ODataBatchReaderStreamScanResult.Match)
+            {
+                int boundaryStartPosition, boundaryEndPosition;
+                scanResult = this.BatchBuffer.ScanForBoundary(
+                    this.CurrentBoundaries,
+                    /*stopAfterIfNotFound*/int.MaxValue,
+                    out boundaryStartPosition,
+                    out boundaryEndPosition,
+                    out isEndBoundary,
+                    out isParentBoundary);
+                switch (scanResult)
+                {
+                    case ODataBatchReaderStreamScanResult.NoMatch:
+                        if (this.underlyingStreamExhausted)
+                        {
+                            // there is nothing else to load from the underlying stream; the requested boundary does not exist
+                            this.BatchBuffer.SkipTo(this.BatchBuffer.CurrentReadPosition + this.BatchBuffer.NumberOfBytesInBuffer);
+                            return false;
+                        }
+
+                        // skip everything in the buffer and refill it from the underlying stream; continue scanning
+                        this.underlyingStreamExhausted = this.BatchBuffer.RefillFrom(this.rawInputContext.Stream, /*preserveFrom*/ODataBatchReaderStreamBuffer.BufferLength);
+

```

Comment:
nit: it would be nice if there was a blank line between break and case for readability.